### ISSUE TITLE
Token stream contains duplicates

### ISSUE THREAD

**Issue by vsajip** [javacc21/javacc21#143]
I found duplicates in the token stream generated by the Python lexer:
```
$ java PLex testfiles/test_path_hook.py  | grep LPAREN
LPAREN: ( 3 34 3 34
LPAREN: ( 14 13 14 13
LPAREN: ( 15 51 15 51
LPAREN: ( 16 17 16 17
LPAREN: ( 17 53 17 53
LPAREN: ( 19 21 19 21
LPAREN: ( 22 24 22 24
LPAREN: ( 22 32 22 32
LPAREN: ( 22 42 22 42
LPAREN: ( 25 1 25 1
LPAREN: ( 25 1 25 1
LPAREN: ( 25 1 25 1
LPAREN: ( 27 20 27 20
LPAREN: ( 31 18 31 18
```
Note the `LPAREN` at (25, 1) appearing three times. I found this problem when debugging the Python-implemented Python lexer and parser.

I also noticed that the Python-implemented Java and JSON parsers and lexers are giving the same results as the Java implementations, but a lot slower than the Python versions before the token-chaining changes I'm working on integrating. I need to investigate further, but I noticed in the Java-implemented Python parser that when setting a breakpoint at line 177 in
https://github.com/javacc21/javacc21/blob/26f94f84fa21d1abe52a54c6b3a5849670a2ae5a/src/ftl/java/Parser.java.ftl#L176-L177
it's called many times with the same `tok`. Is this expected behaviour?

Anyway, where I'm stalled at the moment is that the Python-implemented parser fails on `testfiles/test_path_hook.py` because, it seems, of a parenthesis imbalance leading to `lineJoiningOn` being `true` and causing a `NEWLINE` to be set as unparsed when it shouldn't be (not sure why that doesn't happen in the Java implementation).

**Comment by vsajip** [comment]
> but a lot slower than the Python versions before the token-chaining changes

Actually, ignore that - it might not be the case, still investigating but it might be a debugging artifact.

**Comment by revusky** [comment]
> it's called many times with the same `tok`. Is this expected behaviour?

Yeah, it is the expected behavior, but it does look completely screwy when you're stepping through in a debugger and I guess I should explain why. There are basically two aspects to this:
- The fact that there just tends to be this redundancy (I guess you could call it that) when you scan ahead and then parse. (AFAICT this is unavoidable and just relates to how a tool like this works, so maybe it's not accurate to call it a "redundancy", more like a *perceived* redundancy.)
- The fact that the way activating/deactivating one or more token types causes cached tokens to be thrown away and then they are retokenized.

So, as regards the first point above, suppose you have something like:

      Foobar : "foo" "bar" =>|| "baz" ;

This means that if we have a `Foobar` production at a choice point, we are going to *scan* forward to see if we have a "foo" followed by a "bar" and then if that is the case, we enter the `Foobar` production and we *consume* the "foo" and the "bar"  (and then hopefully a "baz" but no guarantees on that because our scanahead stopped after the "bar"). But you understand, there is a basic conceptual difference between *scanning* tokens and *consuming* them. But this can look redundant from the POV of just stepping through what the lexer does.

So certain things can look very repetitive in a debugger because we *scan* two tokens forward and then we go back and *consume* those two tokens. Now, typically, if we did the scan ahead and we found "foo" followed by "bar", we don't need to retokenize, since the tokens are cached. Before the recent refactoring, they were cached via the tok.next and tok.next.next sort of machinery, but now they are cached the new way but cached nonetheless. But if there was an ACTIVATE_TOKENS there then, the way it works now (and it might be optimizable but I haven't worked on that and don't consider it a priority now) is that it will retokenize in these spots where the set of active tokens changed. (It might not be necessary in given cases, but I have not put any effort into that problem. Not yet, and I might never do so...) So, another way of looking at this is that the code is implemented in a very conservative (and possibly sometimes inefficient way) to ensure maximal correctness. In any case, my point here is to answer the question of why it looks so repetitive when stepping through in a debugger. And the answer is that, even thought it looks screwy, it is basically right. (I'm pretty sure it is. I convinced myself of that, but if you need to convince yourself on your own, by all means... no particular reason to put so much trust in me!)

So you could be running into a lot of what looks very repetitious in particular when you scan/parse the `MatchStatement`. So, like consider what happens with the code here: https://github.com/javacc21/javacc21/blob/master/examples/python/Python.javacc#L475-L479

We *scan* forward until the colon and if that scan succeeds we go back and parse. But you see the way it's written, I guess it needs to retokenize twice, once in the scan pass and then again in the parse pass. This is because of the way ACTIVATE_TOKENS is implemented. See in particular: https://github.com/javacc21/javacc21/blob/master/src/ftl/java/CommonUtils.java.ftl#L157

When we hit that `token_source.reset` method, we're throwing away any cached tokens after the key point to be on the safe side.

I hope that's clear. I have to admit that I found all this quite confusing when implementing and looking at it again, I still find it confusing, but I think it's all correct.

Now, a couple of further observations. First of all, it is quite possible that this bit can be optimized somewhat with an assertion.

For example, the first line of the MatchStatement here: https://github.com/javacc21/javacc21/blob/master/examples/python/Python.javacc#L476

I think (and you can check this) that you can write this more efficiently by adding an assertion at the top:

    MatchStatement : 
        ASSERT {getToken(1).getImage().equals("match")}#
       ACTIVATE_TOKENS MATCH ("match") 
       SubjectExpression 
       ":" 
       =>||

Then the lookahead routine fails if the next token is not a "match". Unless the next token actually is "match" it doesn't need to go to the next line because the lookahead has already failed. But even this has some rather funny aspects that are worth understanding. If the next token is "match", in the ASSERT, it was tokenized as an IDENTIFIER and then goes to the next line and there is a reset and that input is retokenized as a MATCH type because it uncaches the "match" identifier it just tokenized. The implementation is quite intricate in a sense, but I think that from the POV of the application programmer, it actually works reasonably okay -- I mean in terms of principle of least surprise. Well, the point is that, since usually the next token is not "match" the lookahead can just check for that string and fail without hitting the ACTIVATE_TOKENS bit.

Actually, it occurs to me that the maximally efficient assertion there might be:

      ASSERT {currentLookaheadToken==null || getToken(1).getImage().equals("match"}#

You see, as you probably know already, the `#` character at the end of the assertion means that it applies in a lookahead. Without that, it only applies in an actual parsing routine. But actually, we don't need this assertion *unless* we are in a lookahead, so we could just have the assertion always be true when parsing, and the way to do that is to add the currentLookaheadToken==null bit because that variable is only set when we are looking ahead, not when we are parsing. You see, if we entered the production, we know the lookahead succeeded, so we know that the getToken(1).getImage().equals("match") is true, so we don't need to check that again! (Really, just saying, because it probably doesn't make any noticeable difference anyway!)

I'm not confusing you, am I? (Maybe I'm just confusing myself.)

Anyway, nice to see you working on this on Christmas Eve. It made me think of a variant on the classic song: "I'm dreaming of a coding Christmas" or something like that. LOL.

**Comment by revusky** [comment]
> causing a `NEWLINE` to be set as unparsed when it shouldn't be (not sure why that doesn't happen in the Java implementation).

My best guess on this is that this is because of the RESET_TOKEN_HOOK thingy, which I never explained anywhere yet, because I still need to write a blog article (or two) about this stuff. 

I'll try my best to explain...

A RESET_TOKEN_HOOK (and there can be more than one, analogous to the way TOKEN_HOOK works) is basically a hook that allows you to do some housekeeping when the reset(tok) method is called. In this precise case of Python, for example, if we reset the position to some arbitrary point, we need to adjust the `indentationStack` and these variables like `parenthesesNesting` and so on accordingly. That's what this code is about: https://github.com/javacc21/javacc21/blob/master/examples/python/PythonLexer.javacc#L85

The whole thing gets a bit hairy because when reset(tok) is called, `tok` itself can be an INDENT or DEDENT. In that case, it has to set this variable called `danglingIndentation` and that is used to maintain the chain. The key point is here: https://github.com/javacc21/javacc21/blob/master/examples/python/PythonLexer.javacc#L41 and you see that if we have danglingIndentation, i.e. we just reset our position to this INDENT or DEDENT token, we need to prepend that token to the tok passed into the TOKEN_HOOK method. Does that make sense?

I admit that this is very complicated, but I could not find an easier solution. Well, I know it seems overengineered for the problem at hand, but I am looking ahead to the general problem of being able to reset the lexer to any arbitrary token, because I see the need for this to be able to do fault-tolerant or just generally to reset our position in the file.

But it is complex but that's what I did and it does have the redeeming feature that it works. (I think it does...)

Well, a point about all of this is that this machinery is complex for Python, but for many (or most) other languages it may not be such a big deal. In fact, I guess if you are not maintaining any state in the lexer, something analogous to Python's `indentationStack` there may be no need to use the RESET_TOKEN_HOOK thing at all. Or if there is a need for it, it's something much simpler than this.

**Comment by vsajip** [comment]
OK, thanks for the background info - but what about the original point about the repeated token at (25, 1) in that file? Surely that isn't expected behaviour? Or does `PLex.java` need changing somehow?

**Comment by revusky** [comment]
Gee, I just reread your question and I realize that I was answering something quite different from what you asked. I was talking about how the getNextToken is liable to be called more times than you would think because of scanning ahead vs. actual parsing, BUT... you were talking about this PLex test harness which, all it does is lexing, so... well, what I explained is pretty much all irrelevant to your question. (Sorry.)

Offhand, I'm not sure why there are repeated tokens there. I was just eyeballing PLex.java and, of course, I wrote it originally well before this latest refactoring of the token-related code. I also see that you must have edited it since, in particular the tokenizeFile method. My best guess now is that what you're seeing is some artifact of how PLex.java is written, i.e. there is something wrong with it, but that isn't really a problem wrt the Python parser overall. I mean, the fact remains that the overall parsing does work. So when we parse a python source file using PyTest.java, the lexer is surely not putting duplicate tokens in the stream, because the parsing would fail, no?

But anyway (and it's a bit embarassing for me) everything I said above about MatchStatement and so on is actually totally irrelevant to this problem, since if we're just lexing, we never hit any of this logic about activating/deactivating tokens. In fact, if the MATCH and CASE token types are always deactivated, the lexer will just always scan them as IDENTIFIER, but that's not really a problem since we're not parsing anyway.

Well, again, I'm not answering your question. Really, I'd have to look at this more closely and I'm willing to, of course, maybe later today or tomorrow. But I'm also hoping you'll figure it out without me!

**Comment by revusky** [comment]
> Or does `PLex.java` need changing somehow?

To write a short answer for a change... yeah, I think that there must be a problem with PLex.java. But having looked at it just now, I'm not sure exactly what the problem is. I think you can figure it out without me, but if not... well, you know where to find me! :-)

**Comment by vsajip** [comment]
> well, what I explained is pretty much all irrelevant to your question. (Sorry.)

Still good background though, so I'm not complaining :slightly_smiling_face:

**Comment by vsajip** [comment]
> I also see that you must have edited it since, in particular the tokenizeFile method.

That was only a change to add a `quiet` parameter to print tokens conditionally. I also changed it yesterday because after the token-chaining changes, it wasn't terminating but printing `EOF` endlessly in the inner loop
https://github.com/javacc21/javacc21/blob/17e7996a6e9f52b7ab0f030867e08256a3156e3b/examples/python/PLex.java#L78
to which I added an escape clause. Those are the only changes I've made AFAICT. I'll take a closer look at the repeating tokens problem when I get a chance.
